R/clean_data_helpers.R

```diff
@@ -1,54 +1,145 @@
+
+  # save the variable length
+  n_rows <- length(x)
+
+  # get the proportion of NA
+  are_na <- round((sum(is.na(x)) / n_rows), 6L)
+  x      <- x[!is.na(x)]
+
+  # get the proportion of logical values
+  if (is.logical(x)) {
+    are_logical <- round((sum(x) + sum(!x)) / n_rows, 6L)
+    return(c(are_na, are_numeric, are_date, are_character, are_logical))
+  }
+
+  # get the proportion of numeric values
+  l <- levels(x)
+  numeric_levels <- l[!is.na(suppressWarnings(as.numeric(l)))]
+  if (length(numeric_levels) > 0L) {
+    are_numeric <- round((sum(x %in% numeric_levels) / n_rows), 6L)
+  }
+
+  date_levels <- l[!is.na(suppressWarnings(
+    as.Date(
+      lubridate::parse_date_time(
+        l,
+        orders = c("ymd", "ydm", "dmy", "mdy", "myd", "dym", "Ymd", "Ydm",
+                   "dmY", "mdY", "mYd", "dYm")
+      )
+    )
+  ))]
+  if (length(date_levels) > 0L) {
+    are_date <- round((sum(x %in% date_levels) / n_rows), 6L)
+  }
+
+  are_character <- round((1.0 - (are_na + are_numeric +
+                                   are_date + are_logical)), 6L)
+
+  # return the output
+  return(c(are_na, are_numeric, are_date, are_character, are_logical))
+}
+
+
 #' Calculate the percentage of missing and other data type values in a vector
 #' containing different data types such as numeric, Date, character, and
 #' logical.
 #'
 #' @param x A vector of ones or a combination of various data types.
+#' @param type A character with the the vector type.
 #'
 #' @returns A vector of 5 elements representing the percentage of missing,
 #' numeric, date, character, and logical values found in the input vector.
 #'
 #' @keywords internal
 #'
-scan_columns <- function(x) {
-  # --- save the variable length ---
+scan_columns <- function(x, type) {
+  ## The processing path is determine with the logical `verdict`
+  type    <- unlist(strsplit(type, ",", fixed = TRUE))
+  verdict <- type %in% c("logical", "factor")
+
+  ## logical and factor are processed differently from the other.
+  ## Logical vectors are expected to contain `TRUE` or `FALSE` or `NA`. For
+  ## columns of this type, we will check the %NA, %TRUE and %FALSE. The others
+  ## will be set at 0.
+  ## For factor columns, the processing is based on their levels. Levels will be
+  ## checked for numeric, date, logical, NA, and character
+  if (any(verdict)) {
+    return(scan_lgl_and_fct_columns(x))
+  }
+
+  # save the variable length
   n_rows <- length(x)
 
-  # --- get the proportion of NA ---
+  # get the proportion of NA
   are_na <- round((sum(is.na(x)) / n_rows), 6L)
   x      <- x[!is.na(x)]
 
-  # --- get the proportion of numeric values ---
-  tmp         <- suppressWarnings(as.numeric(x))
-  are_numeric <- round((sum(!is.na(tmp)) / n_rows), 6L)
-
-  # --- get the proportion of date values ---
-  x        <- x[which(is.na(tmp))]
+  # get the proportion of date values
   are_date <- 0L
+  if (any(type %in% c("POSIXct", "POSIXt"))) {
```

Comment:
This solution doesn't scale as we can't possibly add special cases for every existing types. Especially since packages can define their own types.